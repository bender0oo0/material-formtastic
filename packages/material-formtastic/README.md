# material-formtastic

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test material-formtastic` to execute the unit tests.
