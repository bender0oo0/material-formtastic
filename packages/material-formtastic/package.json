{
  "name": "material-formtastic",
  "version": "0.0.1",
  "peerDependencies": {
    "@angular/common": "17.0.0",
    "@angular/core": "17.0.0"
  },
  "dependencies": {
    "tslib": "2.3.0"
  },
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "bugs": {
    "url": "https://github.com/bender0oo0/material-formtastic"
  },
  "homepage": "https://github.com/bender0oo0/material-formtastic",
  "repository": {
    "type": "git",
    "url": "https://github.com/bender0oo0/material-formtastic"
  },
  "author": "Marc Frost",
  "license": "MIT",
  "keywords": [
    "angular",
    "angular material",
    "formalar",
    "formalar remote"
  ]
}
