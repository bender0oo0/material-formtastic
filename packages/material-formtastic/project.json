{
  "name": "material-formtastic",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/material-formtastic/src",
  "prefix": "material-formtastic",
  "tags": ["material"],
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/angular:package",
      "outputs": ["{workspaceRoot}/dist/{projectRoot}"],
      "options": {
        "project": "packages/material-formtastic/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "packages/material-formtastic/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "packages/material-formtastic/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production"
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "packages/material-formtastic/jest.config.ts"
      }
    },
    "lint": {
      "executor": "@nx/eslint:lint",
      "outputs": ["{options.outputFile}"]
    }
  }
}
