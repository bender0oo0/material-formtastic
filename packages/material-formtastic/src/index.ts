export * from './lib/material-formtastic/material-formtastic.module';
export * from './lib/material-formtastic/material-formtastic.service';
export * from './lib/material-formtastic/material-formtastic.component';
export * from './lib/material-formtastic/types';
export * from './lib/material-formtastic/abstract-loader.service'
export * from './lib/material-formtastic/http-loader.service';
