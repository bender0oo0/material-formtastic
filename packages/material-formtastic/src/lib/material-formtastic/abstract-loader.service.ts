import {Observable} from "rxjs";
import {FormDefinition, PrimaryKey} from "./types";

export abstract class AbstractLoaderService {
  abstract loadDefinition<T extends object>(primaryKey?: PrimaryKey): Observable<FormDefinition<T>>;
  abstract loadValue<T extends object>(primaryKey: PrimaryKey): Observable<T>;
}

