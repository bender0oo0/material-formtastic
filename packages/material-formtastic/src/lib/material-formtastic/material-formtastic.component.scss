$grid-columns: 24;

.material-formtastic-form {
    display: grid;
    grid-template-columns: repeat($grid-columns, 1fr);
    grid-auto-rows: min-content;
    grid-gap: var(--form-grid-gap);
    margin: var(--form-grid-margin);


    mat-form-field, mat-slide-toggle, mat-checkbox {
      grid-column: 1 / span $grid-columns;
    }

    @for $i from 1 through $grid-columns {
      @for $ii from 1 through $grid-columns {
        .span-#{$i}.start-#{$ii} {
          grid-column: $ii / span $i;
        }
      }
      .span-#{$i} {
        grid-column: auto / span $i;
      }
      .start-#{$i} {
        grid-column-start: $i;
      }
    }
  }
