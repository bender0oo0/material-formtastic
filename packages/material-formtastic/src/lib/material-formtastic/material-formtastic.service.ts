import {Injectable} from '@angular/core';
import {AbstractLoaderService} from "./abstract-loader.service";
import {forkJoin, map, shareReplay} from "rxjs";
import {PrimaryKey} from "./types";

@Injectable()
export class MaterialFormtasticService {
  constructor(private loader: AbstractLoaderService) {
  }

  loadWithoutValue<T extends object>() {
    return this.loader.loadDefinition<T>().pipe(
      shareReplay({ refCount: true, bufferSize: 1 })
    );
  }

  load<T extends object>(primaryKey: PrimaryKey) {
    return forkJoin([this.loader.loadDefinition<T>(primaryKey), this.loader.loadValue<T>(primaryKey)]).pipe(
      map(([def, value]) => {
        def.setValues(value);
        return def;
      }),
      shareReplay({ refCount: true, bufferSize: 1 }),
    );
  }
}
