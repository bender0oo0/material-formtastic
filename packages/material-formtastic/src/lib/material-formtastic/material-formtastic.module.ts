import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {MatIconModule} from "@angular/material/icon";
import {MatCheckboxModule} from "@angular/material/checkbox";
import {MatSlideToggleModule} from "@angular/material/slide-toggle";
import {MatSelectModule} from "@angular/material/select";
import {MatInputModule} from "@angular/material/input";
import {ReactiveFormsModule} from "@angular/forms";
import {MatFormFieldModule} from "@angular/material/form-field";
import {MaterialFormtasticComponent} from "./material-formtastic.component";
import {HttpClientModule} from "@angular/common/http";

@NgModule({
  declarations: [MaterialFormtasticComponent],
  imports: [CommonModule, HttpClientModule, ReactiveFormsModule, MatFormFieldModule, MatInputModule, MatSelectModule, MatSlideToggleModule, MatCheckboxModule, MatIconModule],
  exports: [MaterialFormtasticComponent],
  providers: [],
})
export class MaterialFormtasticModule {
}
