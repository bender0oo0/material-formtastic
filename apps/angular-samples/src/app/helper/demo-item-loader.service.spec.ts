import { DemoItemLoader } from './demo-item-loader.service';
import { PrimaryKey } from 'material-formtastic';

describe('DemoItemLoader', () => {
  let loader: DemoItemLoader;

  beforeEach(() => {
    loader = new DemoItemLoader();
  });

  it('loadValue should return an Observable of item', done => {
    const primaryKey: PrimaryKey = 'test';
    loader.loadValue(primaryKey).subscribe(item => {
      expect(item).toEqual({});
      done();
    });
  });
});
