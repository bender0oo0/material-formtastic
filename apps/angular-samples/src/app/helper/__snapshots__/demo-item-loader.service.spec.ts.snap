// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DemoItemLoader loadValue should return an Observable of item 1`] = `
{
  "color": 3,
  "color2": 1,
  "color3": [
    2,
    3,
  ],
  "confirmed2": true,
  "description": "description 123",
  "description2": "description 123 readonly",
  "description3": "description 123 disabled",
  "enabled2": true,
  "id": 11,
  "id2": 12,
  "id3": 14,
  "name": "bernd",
  "name2": "bernd readonly",
  "name3": "bernd disabled",
}
`;
